src/librustc_metadata/creader.rs

```diff
@@ -764,106 +764,96 @@ impl<'a> CrateLoader<'a> {
-                                                    compiled with `-Z sanitizer`"));
-                            false
-                        }
-                    }
-                }) {
-                    return
-                }
-            }
+        let supported_targets = match *sanitizer {
+            Sanitizer::Address => ASAN_SUPPORTED_TARGETS,
+            Sanitizer::Thread => TSAN_SUPPORTED_TARGETS,
+            Sanitizer::Leak => LSAN_SUPPORTED_TARGETS,
+            Sanitizer::Memory => MSAN_SUPPORTED_TARGETS,
+        };
+        if !supported_targets.contains(&&*self.sess.opts.target_triple.triple()) {
+            self.sess.err(&format!("{:?}Sanitizer only works with the `{}` target",
+                sanitizer,
+                supported_targets.join("` or `")
+            ));
+            return
+        }
 
-            let mut uses_std = false;
-            self.cstore.iter_crate_data(|_, data| {
-                if data.name == sym::std {
-                    uses_std = true;
-                }
-            });
+        let any_non_rlib = self.sess.crate_types.borrow().iter().any(|ct| {
+            *ct != config::CrateType::Rlib
+        });
+        if !any_non_rlib {
+            info!("sanitizer runtime injection skipped, only generating rlib");
+            self.sess.injected_sanitizer_runtime.set(None);
+            return
+        }
 
-            if uses_std {
-                let name = match *sanitizer {
-                    Sanitizer::Address => "rustc_asan",
-                    Sanitizer::Leak => "rustc_lsan",
-                    Sanitizer::Memory => "rustc_msan",
-                    Sanitizer::Thread => "rustc_tsan",
-                };
-                info!("loading sanitizer: {}", name);
-
-                let symbol = Symbol::intern(name);
-                let dep_kind = DepKind::Explicit;
-                let (_, data) =
-                    self.resolve_crate(&None, symbol, symbol, None, None, DUMMY_SP,
-                                       PathKind::Crate, dep_kind)
-                        .unwrap_or_else(|err| err.report());
-
-                // Sanity check the loaded crate to ensure it is indeed a sanitizer runtime
-                if !data.root.sanitizer_runtime {
-                    self.sess.err(&format!("the crate `{}` is not a sanitizer runtime",
-                                           name));
-                }
-            } else {
-                self.sess.err("Must link std to be compiled with `-Z sanitizer`");
+        let any_unsupported = self.sess.crate_types.borrow().iter().any(|ct| {
+            match *ct {
+                config::CrateType::Executable |
+                config::CrateType::Staticlib |
+                config::CrateType::Cdylib |
+                config::CrateType::Dylib |
+                config::CrateType::Rlib => false,
+                _ => true,
+            }
+        });
```

Comment:
Feels like double negation here. Consider rewriting this to mean `all_supported`?